.github/workflows/ci-testing.yaml

```diff
@@ -0,0 +1,258 @@
+#    - Create the Python package
+#    - Upload the Python package for the next job
+# 3. Test-package:
+#    - Download the Python package from the previous job
+#    - Install the downloaded Python package
+#    - Test the code base
+#    - Check if Jupiter Notebooks run without errors
+
+name: Test code, notebooks and package
+
+on:
+  # Trigger the workflow on push
+  push:
+    # Every branch
+    branches:
+      - '**'
+    # But do not run this workflow on creating a new tag starting with 'v', e.g. 'v1.0.3' (see pypi-publish.yml)
+    tags-ignore:
+      - 'v*'
+  # Trigger the workflow on pull request
+  pull_request:
+    branches:
+      - '**'
+  # Allows you to run this workflow manually from the Actions tab
+  workflow_dispatch:
+
+# Allow only one concurrent workflow, skipping runs queued between the run in-progress and latest queued.
+# And cancel in-progress runs.
+concurrency:
+  group:
+    ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
+  cancel-in-progress: true
+
+jobs:
+  # Job 1: Check code quality and consistency
+  code-quality:
+    strategy:
+      matrix:
+        os: [ubuntu-latest]
+        python-version: ['3.12']
+
+    runs-on: ${{ matrix.os }}
+
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v4
+        with:
+          fetch-depth: '0' # full history with tags to get the version number by versioningit
+
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v5
+        with:
+          python-version: ${{ matrix.python-version }}
+
+      - name: Install Python dependencies
+        # Install 'validate-pyproject' for checking pyproject.toml
+        # Install 'ruff' for code linting and formatting
+        # Install 'nbqa' for quality assurance of Jupyter notebooks
+        run: pip install 'validate-pyproject[all]' ruff nbqa
+
+      - name: Install npm dependencies
+        # Install 'prettier' for code formatting of Markdown, YAML, etc. files
+        # Install 'prettier-plugin-toml' plugin for code formatting of TOML files
+        run: npm install prettier prettier-plugin-toml --save-dev --save-exact
+
+      # Check the validity of pyproject.toml
+      - name: Check validity of pyproject.toml
+        id: check_pyproject
+        continue-on-error: true
+        run: validate-pyproject pyproject.toml
+
+      # Check code linting with Ruff in the project root
+      - name: Check code linting
+        id: check_code_linting
+        continue-on-error: true
+        run: ruff check .
+
+      - name: Suggestion to fix code linting issues (for *.py files)
+        if: steps.check_code_linting.outcome == 'failure'
+        run:
+          echo "In project root run 'ruff check . --fix' and commit changes to
+          fix issues."
+
+      # Check code formatting with Ruff in the project root
+      - name: Check code formatting
+        id: check_code_formatting
+        continue-on-error: true
```

Comment:
You are right, we still have to make another commit. But sometimes it happens that several problems occur at the same time, one with code formatting, another with, for example, notebook formatting, etc. I would like to know about all these problems in the first job ('code quality') after one run, because each CI run is not that fast.

Reply:
The first ‘code quality’ job in this workflow now contains more checks such as code linting, code formatting, formatting notepads as well as non-code files. All these checks in this job take only 5-10 seconds to complete. If any of these checks returns an error, the entire first job fails and the next job in this workflow is not executed. Thus, we do not waste resources and time. But the advantage of running all the checks in the first job is that after 5-10 seconds you know whether only one check failed or several. I've had this problem a few times where after fixing code formatting, I re-run CI to see that the markdown or toml file formatting was not done 🙈